[triton] Warp specialization support in torchinductor

mandroid6 opened the pull request:
Summary:
Currently only `num_warps` and `num_stages` are supported as one of the kernel options for inductor auto-tuning using `TritonTemplate`. In order to allow warp-specialization kernel options should allow specifying `num_consumer_groups` and `num_buffers_warp_spec` as well.


Test Plan:
## Unit test
Added tests for `test_triton_template_warp_specialization` to verify generated kenrnel contains configs for  `num_consumer_groups` and `num_buffers_warp_spec`.


## Functional Testing
Specific to flexattention.
```
import torch
from torch.nn.attention.flex_attention import flex_attention

from triton.testing import do_bench

make_tensor = lambda: torch.rand(8, 16, 8192, 128, device="cuda", dtype=torch.bfloat16)
q, k, v = make_tensor(), make_tensor(), make_tensor()

flex_compiled = torch.compile(flex_attention, fullgraph=True)


print(do_bench(lambda: flex_compiled(q, k, v, kernel_options={"num_warps": 4})))
```

triton do_bench results:
- default compile: 15.176783561706543
- with warp-spec: 9.452800750732422

## Extra notes
- generated triton kernel using `TORCH_LOGS=output_code`: P1740612877
- TTGIR for fused kernel: P1740614685

Differential Revision: D70212243




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

eellison commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1881,6 +1881,8 @@ def add_arg(idx, arg, is_constexpr=False, equals_1=False, equals_none=False):
                 "kwargs": config.kwargs,
                 "num_warps": config.num_warps,
                 "num_stages": config.num_stages,
+                "num_consumer_groups": getattr(config, 'num_consumer_groups', 0),
```
Is this only for an internal version of triton ? I don't see these fields as parameters on triton's config: 

https://github.com/triton-lang/triton/blob/18b8d1d0f44cd641b0ac6f041bda5a7d747b9acc/python/triton/runtime/autotuner.py#L251-L269 ?

or do you know where this is documented in triton repro ?

mandroid6 replied:
You are correct, this is only applicable to the facebookexperimental/ws version (https://github.com/facebookexperimental/triton/blob/ws/python/triton/runtime/autotuner.py#L268)

eellison reviewed (commented):
pr looks good but have one clarifying question

htyu commented:
@jansel this is our first attempt to apply warp specialization to inductor Triton templates. Perf is pretty good so far. You feedbacks are much appreciated!

jansel reviewed (changes requested):
failing tests?

Commit:
[pytorch][triton] Warp specialization support in TritonTemplate for torchinductor (#148503)

Summary:

Currently only `num_warps` and `num_stages` are supported as one of the kernel options for inductor auto-tuning using `TritonTemplate`.

In order to allow warp-specialization kernel options should allow specifying `num_consumer_groups` and `num_buffers_warp_spec` as well.


NOTE: Currently gating changes to FBCODE using is_fbcode() checks throughout

Test Plan:
## Unit test
Added tests for `test_triton_template_warp_specialization` to verify generated kenrnel contains configs for  `num_consumer_groups` and `num_buffers_warp_spec`.


## Functional Testing
Specific to flexattention.
```
import torch
from torch.nn.attention.flex_attention import flex_attention

from triton.testing import do_bench

make_tensor = lambda: torch.rand(8, 16, 8192, 128, device="cuda", dtype=torch.bfloat16)
q, k, v = make_tensor(), make_tensor(), make_tensor()

flex_compiled = torch.compile(flex_attention, fullgraph=True)


print(do_bench(lambda: flex_compiled(q, k, v, kernel_options={"num_warps": 4})))
```

triton do_bench results:
- default compile: 15.176783561706543
- with warp-spec: 9.452800750732422

## Extra notes
- generated triton kernel using `TORCH_LOGS=output_code`: P1740612877
- TTGIR for fused kernel: P1740614685

Reviewed By: eellison, devashishshankar, jackiexu1992

Differential Revision: D70212243

mandroid6 commented on torch/_inductor/runtime/autotune_cache.py:
```diff
@@ -464,7 +469,22 @@ def _load_cached_autotuning(
+        }
+
+        # Conditionally add arguments based on inductor_config.is_fbcode()
+        if inductor_config.is_fbcode():
```
> This will work if someone installs the public Meta fork of Triton correct? 

Correct, currently gating this with `is_fbcode` since pytorch-triton is tagged against `triton/release-3.3.x`, whereas the warpspec patches are only available on `triton/release-3.2.x`.

We need to land this to unblock an urgent internal ask; aim is to remove these is_fbcode() changes as soon as triton/release-3.3.x is patched correctly (within 8-9 days).

malfet commented:
@pytorchbot revert -m "Broke lint, see https://hud.pytorch.org/hud/pytorch/pytorch/af7719a2fa0d039c995a7fd1268f627c512bc886/1?per_page=50&name_filter=lint&mergeLF=true" -c ignoredsignal

mandroid6 commented:
Exported in  https://github.com/pytorch/pytorch/pull/150122
